crates/circuit/src/dag_circuit.rs

```diff
@@ -392,6 +393,30 @@ impl PyBitLocations {
     fn __getnewargs__(slf: Bound<Self>) -> PyResult<(Bound<PyAny>, Bound<PyAny>)> {
         Ok((slf.getattr("index")?, slf.getattr("registers")?))
     }
+
+    fn __getitem__<'py>(
+        slf: Bound<'py, Self>,
+        index: PySequenceIndex<'py>,
+    ) -> PyResult<Bound<'py, PyAny>> {
+        let names = [intern!(slf.py(), "index"), intern!(slf.py(), "registers")];
+        if let Ok(index) = index.with_len(2) {
+            match index {
+                crate::slice::SequenceIndex::Int(index) => slf.getattr(names[index]),
+                _ => PyTuple::new(
+                    slf.py(),
+                    index.iter().map(|idx| slf.getattr(names[idx]).unwrap()),
+                )
+                .map(|obj| obj.into_any()),
+            }
+        } else {
+            Err(PyIndexError::new_err("tuple index out of range"))
+        }
```

Comment:
This is a bit more complicated than it needs to be. You don't need to access everything via python attr interface like this, can't you just do the equivalent of something like:

```rust
 if index == 0 {
     return self.index.into_py_any(py);
 } else {
     return self.registers.clone_ref(py).into_any();
  }
```

That also lets you simplify the function signature and just use `&self` and probably elide the lifetimes too.

Reply:
Oh I deliberately chose this over using `&self` to avoid a conversion step.

Reply:
I'm not sure what conversion cost you're worried about here. Using &self is the equivalent of basically just doing `.borrow()` to get `&Self`. The way this is written write not has a lot of extra overhead because you go back to python for every attribute access when it's faster to just get the fields from the rust struct directly. There is a conversion from the rust integer to a python integer but you have that either way.

Reply:
Got it, I have addressed your comments in c315871